Issue with registry configuration for npm.pkg.github.com

Any progress on this please? Haven't been able to use `pnpm` because of this for a long time. My situation:
`.npmrc`
```
@apify-packages:registry=https://npm.pkg.github.com/apify-packages
//npm.pkg.github.com/:_authToken=<token>
```
`pnpm i`
```
 ERR_PNPM_FETCH_401  GET https://npm.pkg.github.com/apify-packages/@apify-packages%2Ffingerprint-injector: Unauthorized - 401

No authorization header was set for the request.

These authorization settings were found:
@apify-packages:registry=https://npm.pkg.github.com/apify-packages
//npm.pkg.github.com/:_authToken=72d4[hidden]
Progress: resolved 1, reused 0, downloaded 0, added 0
```
Any advice would be highly appreciated.

_Originally posted by @cybairfly in https://github.com/pnpm/pnpm/issues/2933#issuecomment-919991675_

--- a/.changeset/forty-yaks-cheat.md
+++ b/.changeset/forty-yaks-cheat.md
@@ -1,0 +1,10 @@
+---
+"@pnpm/config": patch
+"@pnpm/normalize-registries": patch
+---
+
+When normalizing registry URLs, a trailing slash should only be added if the registry URL has no path.
+
+So `https://registry.npmjs.org` is changed to `https://registry.npmjs.org/` but `https://npm.pkg.github.com/owner` is unchanged.
+
+Related issue: [#4034](https://github.com/pnpm/pnpm/issues/4034).

--- a/.changeset/silent-kangaroos-whisper.md
+++ b/.changeset/silent-kangaroos-whisper.md
@@ -1,0 +1,12 @@
+---
+"pnpm": patch
+---
+
+pnpm should read the auth token of a github-registry-hosted package, when the registry path contains the owner [#4034](https://github.com/pnpm/pnpm/issues/4034).
+
+So this should work:
+
+```
+@owner:registry=https://npm.pkg.github.com/owner
+//npm.pkg.github.com/:_authToken=<token>
+```

--- a/packages/config/package.json
+++ b/packages/config/package.json
@@ -41,6 +41,7 @@
     "camelcase": "^6.2.0",
     "can-write-to-dir": "^1.1.1",
     "is-subdir": "^1.1.1",
+    "normalize-registry-url": "2.0.0",
     "ramda": "^0.27.1",
     "realpath-missing": "^1.1.0",
     "which": "^2.0.2"

--- a/packages/config/src/getScopeRegistries.ts
+++ b/packages/config/src/getScopeRegistries.ts
@@ -1,13 +1,11 @@
+import normalizeRegistryUrl from 'normalize-registry-url'
+
 export default function getScopeRegistries (rawConfig: Object) {
   const registries = {}
   for (const configKey of Object.keys(rawConfig)) {
     if (configKey[0] === '@' && configKey.endsWith(':registry')) {
-      registries[configKey.substr(0, configKey.indexOf(':'))] = normalizeRegistry(rawConfig[configKey])
+      registries[configKey.substr(0, configKey.indexOf(':'))] = normalizeRegistryUrl(rawConfig[configKey])
     }
   }
   return registries
-}
-
-export function normalizeRegistry (registry: string) {
-  return registry.endsWith('/') ? registry : `${registry}/`
 }

--- a/packages/config/src/index.ts
+++ b/packages/config/src/index.ts
@@ -9,10 +9,11 @@ import camelcase from 'camelcase'
 import loadNpmConf from '@zkochan/npm-conf'
 import npmTypes from '@zkochan/npm-conf/lib/types'
 import { sync as canWriteToDir } from 'can-write-to-dir'
+import normalizeRegistryUrl from 'normalize-registry-url'
 import fromPairs from 'ramda/src/fromPairs'
 import realpathMissing from 'realpath-missing'
 import whichcb from 'which'
-import getScopeRegistries, { normalizeRegistry } from './getScopeRegistries'
+import getScopeRegistries from './getScopeRegistries'
 import findBestGlobalPrefix from './findBestGlobalPrefix'
 import { getCacheDir, getConfigDir, getDataDir, getStateDir } from './dirs'
 import {
@@ -249,7 +250,7 @@ export default async (
     { 'user-agent': pnpmConfig.userAgent },
   ] as any) // eslint-disable-line @typescript-eslint/no-explicit-any
   pnpmConfig.registries = {
-    default: normalizeRegistry(pnpmConfig.rawConfig.registry),
+    default: normalizeRegistryUrl(pnpmConfig.rawConfig.registry),
     ...getScopeRegistries(pnpmConfig.rawConfig),
   }
   pnpmConfig.lockfileDir = pnpmConfig.lockfileDir ?? pnpmConfig.lockfileDirectory ?? pnpmConfig.shrinkwrapDirectory

--- a/packages/config/test/index.ts
+++ b/packages/config/test/index.ts
@@ -178,7 +178,7 @@ test('when using --global, link-workspace-packages, shared-workspace-shrinwrap a
   }
 })
 
-test('registries of scoped packages are read', async () => {
+test('registries of scoped packages are read and normalized', async () => {
   const { config } = await getConfig({
     cliOptions: {
       userconfig: path.join(__dirname, 'scoped-registries.ini'),
@@ -193,6 +193,7 @@ test('registries of scoped packages are read', async () => {
     default: 'https://default.com/',
     '@foo': 'https://foo.com/',
     '@bar': 'https://bar.com/',
+    '@qar': 'https://qar.com/qar',
   })
 })
 
@@ -215,6 +216,7 @@ test('registries in current directory\'s .npmrc have bigger priority then global
     default: 'https://pnpm.io/',
     '@foo': 'https://foo.com/',
     '@bar': 'https://bar.com/',
+    '@qar': 'https://qar.com/qar',
   })
 })
 

--- a/packages/config/test/scoped-registries.ini
+++ b/packages/config/test/scoped-registries.ini
@@ -1,3 +1,4 @@
 @foo:registry=https://foo.com
 @bar:registry=https://bar.com
+@qar:registry=https://qar.com/qar
 registry=https://default.com

--- a/packages/lockfile-utils/package.json
+++ b/packages/lockfile-utils/package.json
@@ -42,7 +42,7 @@
     "@pnpm/resolver-base": "workspace:8.1.1",
     "@pnpm/types": "workspace:7.6.0",
     "dependency-path": "workspace:8.0.6",
-    "get-npm-tarball-url": "^2.0.2",
+    "get-npm-tarball-url": "^2.0.3",
     "ramda": "^0.27.1"
   },
   "funding": "https://opencollective.com/pnpm"

--- a/packages/normalize-registries/package.json
+++ b/packages/normalize-registries/package.json
@@ -27,7 +27,7 @@
   },
   "dependencies": {
     "@pnpm/types": "workspace:7.6.0",
-    "normalize-registry-url": "1.0.0"
+    "normalize-registry-url": "2.0.0"
   },
   "homepage": "https://github.com/pnpm/pnpm/blob/master/packages/normalize-registries#readme",
   "funding": "https://opencollective.com/pnpm"

--- a/packages/resolve-dependencies/package.json
+++ b/packages/resolve-dependencies/package.json
@@ -44,7 +44,7 @@
     "@pnpm/types": "workspace:7.6.0",
     "dependency-path": "workspace:8.0.6",
     "encode-registry": "^3.0.0",
-    "get-npm-tarball-url": "^2.0.2",
+    "get-npm-tarball-url": "^2.0.3",
     "path-exists": "^4.0.0",
     "ramda": "^0.27.1",
     "replace-string": "^3.1.0",

